[Pytorch Profiler] Add support for adding module hierarchy to

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62419
* #62418
* #62417
* #62228
* #62191
* __->__ #61792

KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)

kimishpatel committed:
[Pytorch Profiler] Add support for adding module hierarchy to
KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

raziel commented on torch/autograd/profiler.py:
```diff
@@ -358,6 +358,10 @@ class profile(object):
         with_stack (bool, optional): record source information (file and line number) for the ops.
 
+        with_module_hierarchy (bool): record module hierarchy (including function names)
+            corresponding to the callstack of the op. e.g. If A's forward call's B's forward
+            and it has aten::add op then aten::add's module hierarchy is A.B
```
```suggestion
            corresponding to the callstack of the op. e.g. If module A's forward calls module B's forward
            which contains an aten::add op, then aten::add's module hierarchy is A.B
```

raziel commented on torch/csrc/autograd/profiler_kineto.h:
```diff
@@ -203,6 +217,7 @@ struct TORCH_API KinetoEvent {
   uint8_t activity_type_ = 0;
   c10::optional<std::vector<std::vector<int64_t>>> shapes_;
   c10::optional<std::vector<std::string>> stack_;
+  c10::optional<std::string> module_hierarchy_;
```
Do you see any future need to parse this hierarchy that calls out for having the hierarchy broken into its different levels? Basically just wondering if this should be a vector as well.
Just wondering...

kimishpatel replied:
@raziel probably yes. I can imagine that we build a tool integrated with the profiler that provides hierarchical break down of latency, but that would require some post processing anyway (such as modules marked with "SELF" are the same module that was seen a level up in hierarchy).

raziel commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -777,6 +777,79 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
  public:
+  std::string moduleHierarchy() const {
+    std::string module_hierarchy("TOP");
+    std::vector<StackEntry> entries;
```
entries is unused

raziel commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -777,6 +777,79 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
+          g_self_type = g_self_type.substr(g_self_type.find_last_of('.') + 1);
+        }
+      }
+      module_hierarchy += "(" + g_self_type + ")::" + fn_name;
```
not be a bad suggestion to use 'operator+=' or 'string::append()'

kimishpatel replied:
Any particular reason as to why one vs. the other? Is it to avoid intermediate string construction?

raziel replied:
yep that's right, which is particularly relevant in the interpreter code.

raziel commented on torch/profiler/profiler.py:
```diff
@@ -116,6 +116,9 @@ class profile(object):
         with_flops (bool): use formula to estimate the FLOPS of specific operators
             (matrix multiplication and 2D convolution).
+        with_module_hierarchy (bool): record module hierarchy (including function names)
+            corresponding to the callstack of the op. e.g. If A's forward call's B's forward
+            and it has aten::add op then aten::add's module hierarchy is A.B
```
```suggestion
            corresponding to the callstack of the op. e.g. If module A's forward calls module B's forward
            which contains an aten::add op, then aten::add's module hierarchy is A.B
```

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

raziel reviewed (commented):
Looks good to me but definitely the others should review and likely have relevant comments.

gdankel commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -268,6 +274,10 @@ void pushProfilingCallbacks() {
         }
+        if (config.with_module_hierarchy &&
+            fn.scope() != at::RecordScope::BACKWARD_FUNCTION) {
+          ctx_ptr->module_hierarchy = jit::currentModuleHierarchy();
```
Can the main part of the work actually be moved to finalizeCPUTrace()?

gdankel replied:
Why the check for backward function here and above?

kimishpatel replied:
> Why the check for backward function here and above?

Good question. For now I just follow other pattern for stacktrace. I am not 100% sure if same information is available for backward functions.

kimishpatel replied:
> Can the main part of the work actually be moved to finalizeCPUTrace()?

Let me check. It has same constraints as `jit::currentStackTrace`. By the time we go to finalizeCPUTrace scope of JIT is gone and we wont get any information.

For the purpose of reducing overhead, we may have to do it differently, but I havent yet thought about it.

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

raziel commented on torch/csrc/autograd/init.cpp:
```diff
@@ -73,7 +73,13 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
+      .def(py::init<ProfilerState,
+          bool, /* record_input_shapes */
+          bool, /* profile_memory */
+          bool, /* with_stac k*/
```
```suggestion
          bool, /* with_stack*/
```

raziel commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -776,6 +777,95 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
   }
 
  public:
+  // One way to avoid overhead of forming string would be to return
```
Currently this returns the hierarchy as a vector of module definitions as mangled <type>::<function_name> (plus some special keys for things like top modules, etc.).

I suggest documenting this as well (i.e. what this function actually returns).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Profiler] Add support for adding module hierarchy to"


KinetoEvent

Summary:
This PR adds module hierarchy information to events.
What is module hierarchy information attached to events?
During profiling a TorchScript module, when events are added, we ask JIT
what is the module hierarchy associated with the node being
executed. At the time of execution of that node, there might be multiple
frames in the stack of interpreter. For each frame, we find
corresponding node and the corresponding module hierarchy is queried.
Module hierarchy corresponding to the node is associated with node's
InlinedCallStack. InlinedCallStack of node tracks the path via which the
node is inlined. Thus during the inlining process we annotate
module information corresponding to the CallMethod nodes being inlined.

With this PR, chrome trace will contain additional metadata:
"Module Hierarchy". This can look like this:
TOP(ResNet)::forward.SELF(ResNet)::_forward_impl.layer1(Sequential)::forward.0(BasicBlock)::forward.conv1(Conv2d)::forward.SELF(Conv2d)::_conv_forward
It contains module instance, type name and the method name in the
callstack.

Test Plan:
test_profiler

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29745442](https://our.internmc.facebook.com/intern/diff/D29745442)


![Screen Shot 2021-07-16 at 4 59 05 PM](https://user-images.githubusercontent.com/17488981/126018845-f00dab93-1172-410c-a022-fd44f9e8e54c.png)


[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29745442).
